File: 2021/3/partOne.ts
import { readFileSync } from "fs";

async function main() {
  let result = 0;
  const file = readFileSync(process.argv[2], "utf-8");
  const rows = file.split("\n").map((r) => {
    const row = r;
    return r;
  });

  let gammaRate = "";
  let epsilonRate = "";

  const amount = rows.length;
  const l = rows[0].length;

  for (let i = 0; i < l; i++) {
    let sum = 0;
    rows.forEach((r, index) => {
      sum += parseInt(r.charAt(i));
    });

    if (sum > amount / 2) {
      gammaRate = `${gammaRate}1`;
      epsilonRate = `${epsilonRate}0`;
    } else {
      gammaRate = `${gammaRate}0`;
      epsilonRate = `${epsilonRate}1`;
    }
  }

  result = parseInt(gammaRate, 2) * parseInt(epsilonRate, 2);

  console.log(result);
}

main();

File: 2021/3/part-one.py
import sys

if __name__ == "__main__":
    result = 0
    with open(sys.argv[1], encoding="utf-8") as f:
        rows = [x.rstrip() for x in f]

    amount = len(rows)
    gamma_rate, epsilon_rate = "", ""

    for i in range(0, len(rows[0])):
        sum = 0
        for r in rows:
            sum += int(r[i])

        gamma_rate += "1" if sum > amount / 2 else "0"
        epsilon_rate += "0" if sum > amount / 2 else "1"

    result = int(gamma_rate, 2) * int(epsilon_rate, 2)

    print(result)
